modules/minifact/plans/retrieve.pp

```diff
@@ -0,0 +1,53 @@
+# A plan that retrieves facts from the specified nodes by running
+# an apropriate minifact::* task on each. Care is taken to run a
+# minifact::* task corresponding to the node's platfrom on each
+# node. If there is no variant of the task appropriate for some
+# node's platform the node is treated as if the task had failed.
+#
+# The $nodes parameter is a list of the nodes to retrieve the facts
+# from.
+plan minifact::retrieve(TargetSpec $nodes) {
+  $targets = get_targets($nodes)
+
+  # Build a mapping from the names of the tasks to run to the lists of
+  # targets to run the tasks on
+  $task_targets = $targets.reduce({}) |$task_targets, $target| {
```

Comment:
This is kind of unreadable. We could simplify this by leveraging Ruby's `group_by` in a separate function. Then it would be
```
$task_targets = $targets.group_by |$target| {
  $target.protocol ? {
    'ssh'   => 'minifact::bash',
    'winrm' => 'minifact::powershell',
    'local' => if minifact::is_command_available('bash') { 'minifact::bash' } else { undef },
    default => undef,
  }
}
```

I guess we could also break it out in a separate Puppet function that re-implements `group_by` as you have here.

Reply:
Also, there's no support for pcp (which could directly use facter, i.e. the facter_task). Was that a separate ticket?

Reply:
The `group_by` function seems too general to include in the `minifact` module, it would be better suited for the `stdlib` module IMO. So I'm going to add the puppet function.

Reply:
I'm going to add the pcp support through a ruby task which should run on all platforms supported by puppet agent.

Reply:
It seems it's not possible to write a native puppet function accepting a block parameter so I ended up writing a ruby based function instead. Since it is namespaced, we don't risk a conflict with a globally defined `group_by` function (if/when added).